tensorflow/core/kernels/data/sampling_dataset_op.cc

```diff
@@ -0,0 +1,202 @@
+                             IteratorStateReader* reader) override {
+        return Status::OK();
+      }
+    };
+
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {
+        std::random_device device("/dev/urandom");
+        std::mt19937_64 rng_temp(device());
+        rng = rng_temp;
+      }
+
+      Status Initialize(IteratorContext* ctx) override {
+        return dataset()->input_->MakeIterator(ctx, prefix(), &input_impl_);
+      }
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        // NOTE(mrry): This method is thread-safe as long as
+        // `input_impl_` and `f` are thread-safe. However, if multiple
+        // threads enter this method, outputs may be observed in a
+        // non-deterministic order.
+        bool rand_val_hit;
+        do {
+          {
+            tf_shared_lock l(mu_);
+            if (!input_impl_) {
+              *end_of_sequence = true;
+              return Status::OK();
+            }
+            TF_RETURN_IF_ERROR(
+                input_impl_->GetNext(ctx, out_tensors, end_of_sequence));
+          }
+          if (*end_of_sequence) {
+            mutex_lock l(mu_);
+            input_impl_.reset();
+            return Status::OK();
+          }
+
+          // generate a number from random uniform [0, 1)
+          float rand_val = unif(rng);
+          rand_val_hit = rand_val < dataset()->rate_;
+          if (!rand_val_hit) {
+            // Clear the output tensor list since it doesn't match.
+            out_tensors->clear();
+          }
+        } while (!rand_val_hit);
+        *end_of_sequence = false;
+        return Status::OK();
+      }
+
+     protected:
+      Status SaveInternal(IteratorStateWriter* writer) override {
+        mutex_lock l(mu_);
+        if (input_impl_) {
+          TF_RETURN_IF_ERROR(SaveInput(writer, input_impl_));
+        } else {
+          TF_RETURN_IF_ERROR(
+              writer->WriteScalar(full_name("input_impl_empty"), ""));
+        }
+        return Status::OK();
+      }
+
+      Status RestoreInternal(IteratorContext* ctx,
+                             IteratorStateReader* reader) override {
+        mutex_lock l(mu_);
+        if (!reader->Contains(full_name("input_impl_empty"))) {
+          TF_RETURN_IF_ERROR(RestoreInput(ctx, reader, input_impl_));
+        } else {
+          input_impl_.reset();
+        }
+        return Status::OK();
+      }
+
+     private:
+      mutex mu_;
+      std::unique_ptr<IteratorBase> input_impl_ GUARDED_BY(mu_);
+      std::mt19937_64 rng;
+      std::uniform_real_distribution<float> unif;
+    };
+
+    const float rate_;
+    const DatasetBase* const input_;
+  };
+};
+
+REGISTER_KERNEL_BUILDER(Name("SamplingDataset").Device(DEVICE_CPU),
```

Comment:
The op name should be `ExperimentalSamplingDataset`.

Reply:
Fixed in https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-a87bcfeaacbb9697bc5998804322ce6cR218